File: 백준 문제 풀이 모음/python/class 4/1932.py
import sys

n = int(sys.stdin.readline()) 
result = []
for i in range(n) :
    tmp = list(map(int, sys.stdin.readline().split()))

    if i == 0 :
        result = tmp
    
    else :
        length = len(tmp)

        for j in range(length) :
            if j == 0 :
                tmp[j] = tmp[j] + result[0] 

            elif j == length -1 :
                tmp[j] = tmp[j] + result[-1]
            
            else :

                tmp[j] = max(tmp[j] + result[j], tmp[j] + result[j-1])
        
        result = tmp

print(max(result))

File: 백준 문제 풀이 모음/kotlin/class 4/1932.kt

private lateinit var matrix : MutableList<MutableList<Int>>

fun main() = with(System.`in`.bufferedReader()){
    val n = readLine().toInt()

    matrix = MutableList<MutableList<Int>>(n+1) { MutableList<Int> (0) { 0 } }

    for (i in 1 .. n) {
        val input = readLine().split(' ').toList().map { it.toInt() }
        matrix[i] = input.toMutableList()
    }

//    matrix.forEach {
//        println(it)
//    }
//

    for (i in 2 until n+1) {
        val arr_size = matrix[i].size
//        println(arr_size)
        for (j in 0 until arr_size) {
            if (j == 0) {
                matrix[i][j] += matrix[i-1][j]
            } else if (j == arr_size -1) {
                matrix[i][j] += matrix[i-1][j-1]
            } else {
                matrix[i][j] += Math.max(matrix[i-1][j-1], matrix[i-1][j])
            }
        }

    }
    print(matrix[n].max())

//
//    matrix.forEach {
//        println(it)
//    }

}

